Added docs for credentials (#56)
danielpaul/RapidRails: docs/credentials_and_secrets.md

# Credentials and Local Secrets

> We use [Rails credentials]((https://edgeguides.rubyonrails.org/security.html#environmental-security) for storing confidential data like API keys, database passwords..etc. It is used when the keys are shared by all the devs and we want it to exist in our codebase in version control.

Reviewer: ```suggestion
We use [Rails credentials](https://edgeguides.rubyonrails.org/security.html#environmental-security) for storing confidential data like API keys, database passwords..etc. It is used when the keys are shared by all the devs and we want it to exist in our codebase in version control.
```

Local secrets ([environment variables](https://guides.rubyonrails.org/v5.1/configuring.html#rails-environment-settings)) are used when devs might have different keys for their local setup or when there are multiple servers in different environments on the cloud platform(like Heroku).

> ## Credentials

Reviewer: ```suggestion
## Rails Credentials File
```

- If `bin/setup` was run previously, a `master.key` should already exist and you can skip the next steps.
- If your admin has already setup the key, create a `master.key` file in the config folder and paste the key in the file without any line breaks.

> - If both the above steps don't apply and a new key needs to be generated, run `EDITOR='echo' bin/rails credentials:edit` to create a new one.

Reviewer: we can skip about the bin/setup command here since we add it at the start. this is all about manual setup. 

need to say if it's a new app, then they would delete the credentials file and then run the command to create a new one. 

if its an existing project then they would get the master.key from the other dev to decrypt the file that they have.

### Viewing the credentials file

To view the application credentials using VS code, run `EDITOR='code --wait' rails credentials:edit` in your terminal. Replace `code` in the above command with the command of your specific editor if you're using another IDE.

### Credentials file setup

We use a single credentials file for all environments which is located at [config/credentials.yml.enc](../config/credentials.yml.enc).

> For more details on Rails credentials and how to access or write to the file, refer to this [documentation](https://edgeguides.rubyonrails.org/security.html#environmental-security)

Reviewer: this can be the first paragraph in the section. and we don't need this to be in the subheading here.

## Local Secrets

### Local Secrets setup on local system

We use [dotenv-rails](https://github.com/bkeepers/dotenv) to store local secrets on your local system.

1. Duplicate the `.env.template` file and set the filename as `.env`. You can skip this step if you've run the `bin/setup` command previously.
2. You can store your local secrets here like `HOST=localhost:3000`.
3. To access it anywhere in the application, use the `ENV` object on the variable like `ENV['HOST']`.

### Local Secrets setup on cloud platform

> 1. On your cloud platform, set the same value for `HOST` in your **config variables** settings.

Reviewer: add a screenshot of how this is on Heroku here. also link to Heroku's docs on adding env variables. you can get the image from there also id say.

> 2. The value of these variable can be quickly changes in the settings without re-deploying the app.

Reviewer: ```suggestion
2. The value of these variable can be quickly changed without re-deploying the app.
```

# Credentials and Local Secrets

Rails credentials allows you to securely store sensitive configuration information and credentials for your Rails application. This feature is particularly useful for storing things like API keys, database passwords, or any other confidential data that your application needs to access.

> Local secrets (environment variables) are secrets that are environment-specific information unique to your local system or the cloud platform you use such as Heroku. We use these variables when every local system requires different credentials. They can also be used to quickly change certain settings that might change very often like disabling a certain feature temporarily. There is no need to update any code and re-deploy the app in this case.

Reviewer: link to the official documentation on how credentials file works. 

in here we just need to say how we use that credentials file and also .env file for local variables. 

talk about the difference for when we would use either. 
* credential files - when the same keys needs to be shared by all devs and want keys to be in version control
* .env file variables that can be access with ENV[''] method - env specific - both for each developer in their machines or if they have multiple deployed servers, server specific keys

## Credentials

> We use the default rails credentials file to store application credentials that can be accessed only by the app with the requirement of having a `master.key` file. This file will reside only on your local and your cloud platform as a config variable named `RAILS_MASTER_KEY`. The `master.key` file should not be pushed to github for security purposes.

Reviewer: this should be a given here. but talk about how the setup will help you set the key up when onboarding. 

if they were to do this manually they would need to create the key file. 
* use the same value for the app if this is not a new project so they can decrypt the existing file
* otherwise how to generate a new credentials file replacing the current one for a new project. (give manual instructions for the command we run in the setup but how they can do that manually without running the setup script)

### Viewing the credentials file

To view the application credentials using VS code, run `EDITOR='code --wait' rails credentials:edit` in your terminal. Replace `code` in the above command with the command of your specific editor if you're using another IDE.

> ### Credentials file setup

Reviewer: we don't need instructions on this. they can have a look themselves. we can link to the file directly in our codebase. they can follow the link to then view the sample file then.

The credentials file will look like so. Credentials have different sections for each environment. Environments can also inherit from the **default** credentials. For example, **development** specific credentials will be nested under the `development:` section.

```
secret_key_base: xxx
default: &default
  global:
    key: 1234

test:
  <<: *default

development:
  <<: *default

  test:
  	key: abcdefg

staging:
  <<: *default

production:
  <<: *default

  postmark_api_token: XXX

  aws:
    access_key_id: XXX
    secret_access_key: XXX
    region: eu-west-2
    bucket: rapid_rails_production

  forest_admin:
    env_secret: XXX
    auth_secret: XXX
```

- To read the value specific key within the app, run a command like so:
  `Rails.application.credentials.dig(Rails.env.to_sym, :test, :key)`
  The first param is the rails environment as a symbol. The params are the nested keys for the credentials in the order of nesting. The above command in a **development** environment would return **abcdefg** from the sample credentials file above. You may replace `Rails.env.to_sym` with the specific environment like `:production` to access only production credentials ifs required.

> - Running`Rails.application.credentials.dig(Rails.env.to_sym, :global, :key)` in any environment will return **1234** as all the environments inherit the **default** credentials.

Reviewer: don't need these instructions since we will be linking to the official docs on how this works

## Local Secrets

### Local Secrets setup on local system

We use [dotenv-rails](https://github.com/bkeepers/dotenv) to store local secrets on your local system.

> 1. Duplicate the `.env.template` file and set the filename as `.env`.

Reviewer: mention that they don't need to do this if they run the setup script which will set up things for them
